apps/workers/sql-load/worker.go

```diff
@@ -304,32 +324,62 @@ func (w *worker) QuerySchema() (err error) {
 func (ds *DataSet) ReadFiles(ctx context.Context, files file.Reader, rowChan chan Row, skipErrors bool) {
 	errChan := make(chan error, 20)
 	dataIn := make(chan []byte, 20)
+	var header []string
+	var hBytes []byte
 	var activeThreads int32
+
 	for i := 0; i < 20; i++ {
 		activeThreads++
-		go func() { // unmarshaler
+		go func() { // csv data parsing
 			defer func() { atomic.AddInt32(&activeThreads, -1) }()
 			for b := range dataIn {
-				var j Jsondata
-				if e := json.Unmarshal(b, &j); e != nil {
-					errChan <- fmt.Errorf("json unmarshal error %w %q", e, string(b))
-					return
-				}
-
-				if row, err := MakeRow(ds.dbSchema, j); err != nil {
-					errChan <- fmt.Errorf("%w", err)
-				} else if row != nil {
-					atomic.AddInt32(&ds.rowCount, 1)
-					rowChan <- row
+				if ds.csv { // csv data parsing
+					if bytes.Equal(b, hBytes) {
+						continue // another header row was found skip
+					}
+					if row, e := MakeCsvRow(ds.dbSchema, b, header, ds.delimiter); e != nil {
+						errChan <- fmt.Errorf("csv read error %w %q", e, string(b))
+					} else if row != nil {
+						atomic.AddInt32(&ds.rowCount, 1)
+						rowChan <- row
+					}
+				} else { // json data parsing
+					var j JsonData
+					if e := json.Unmarshal(b, &j); e != nil {
+						errChan <- fmt.Errorf("json unmarshal error %w %q", e, string(b))
+						return
+					}
+
+					if row, err := MakeRow(ds.dbSchema, j); err != nil {
+						errChan <- fmt.Errorf("%w", err)
+					} else if row != nil {
+						atomic.AddInt32(&ds.rowCount, 1)
+						rowChan <- row
+					}
 				}
 			}
 		}()
 	}
 
 	// read the lines of the file
+	h := true // csv header data needs to be set
 	scanner := file.NewScanner(files)
 loop:
 	for scanner.Scan() {
+		// read the first data bytes to capture the header for csv data
+		if ds.csv && h {
+			var err error
+			hBytes = scanner.Bytes()
+			header, err = MakeCsvHeader(hBytes, ds.delimiter)
+			if err != nil {
+				ds.err = fmt.Errorf("csv header read error %w %q", err, string(hBytes))
+				return
```

Comment:
you want a `break loop` instead of a return. The return will leave the processing threads running and won't properly cleanup the file.